wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+		self::WINDOW_SYMBOL                          => 80000,
+		self::ZONE_SYMBOL                            => 80022,
+	);
+
+	/**
+	 * Identifier-like strings that may represent underscore-prefixed charset names.
+	 *
+	 * Includes charsets from both MySQL 5 and 8; via "SHOW CHARACTER SET"/docs:
+	 *   https://dev.mysql.com/doc/refman/5.7/en/charset-charsets.html
+	 *   https://dev.mysql.com/doc/refman/8.4/en/charset-charsets.html
+	 */
+	const UNDERSCORE_CHARSETS = array(
+		'_armscii8' => true,
+		'_ascii'    => true,
+		'_big5'     => true,
+		'_binary'   => true,
+		'_cp1250'   => true,
+		'_cp1251'   => true,
+		'_cp1256'   => true,
+		'_cp1257'   => true,
+		'_cp850'    => true,
+		'_cp852'    => true,
+		'_cp866'    => true,
+		'_cp932'    => true,
+		'_dec8'     => true,
+		'_eucjpms'  => true,
+		'_euckr'    => true,
+		'_gb18030'  => true,
+		'_gb2312'   => true,
+		'_gbk'      => true,
+		'_geostd8'  => true,
+		'_greek'    => true,
+		'_hebrew'   => true,
+		'_hp8'      => true,
+		'_keybcs2'  => true,
+		'_koi8r'    => true,
+		'_koi8u'    => true,
+		'_latin1'   => true,
+		'_latin2'   => true,
+		'_latin5'   => true,
+		'_latin7'   => true,
+		'_macce'    => true,
+		'_macroman' => true,
+		'_sjis'     => true,
+		'_swe7'     => true,
+		'_tis620'   => true,
+		'_ucs2'     => true,
+		'_ujis'     => true,
+		'_utf16'    => true,
+		'_utf16le'  => true,
+		'_utf32'    => true,
+		'_utf8'     => true,
+		'_utf8mb3'  => true,
+		'_utf8mb4'  => true,
+	);
+
+	protected $input;
+	protected $c; // Current character.
+	protected $n; // Next character.
+	protected $position = 0;
+	protected $token;
+	protected $text    = '';
+	protected $channel = self::CHANNEL_DEFAULT;
+	public $type;
+	protected $token_instance;
+	protected $server_version;
+	protected $sql_modes;
+	protected $in_version_comment = false;
+
+	public function __construct( string $input, int $server_version = 80038, int $sql_modes = 0 ) {
+		$this->input          = $input;
+		$this->server_version = $server_version;
+		$this->sql_modes      = $sql_modes;
+		$this->c              = $this->input[ $this->position ] ?? null;
+		$this->n              = $this->input[ $this->position + 1 ] ?? null;
+	}
+
+	public function is_sql_mode_active( int $mode ): bool {
+		return ( $this->sql_modes & $mode ) !== 0;
+	}
+
+	public function get_server_version() {
+		return $this->server_version;
+	}
+
+	public static function get_token_id( string $token_name ): int {
+		if ( 'ε' === $token_name ) {
+			return self::EMPTY_TOKEN;
+		}
+		return constant( self::class . '::' . $token_name );
+	}
+
+	public static function get_token_name( $token_type ): string {
+		$reflection = new ReflectionClass( self::class );
+		$constants  = array_reverse( $reflection->getConstants() ); // some constants can conflict, tokens are at the end
+		$token_name = array_search( $token_type, $constants, true );
+		return $token_name ? $token_name : '<INVALID>';
```

Comment:
That's a lot of computations every time we want a token name. Is this mostly used for debugging? If so, I guess that's fine.

Reply:
@adamziel It's only for debugging indeed. Unless one calls `$token->get_name()` or `(string)$token`, it won't be called.